File: src/Sample/Plumix.Sample/Demos/Material/SearchDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/material/search_demo_page.dart

public sealed class SearchDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new SearchDemoPageState();
    }
}

internal sealed class SearchDemoPageState : State
{
    private static readonly IReadOnlyList<string> SearchTerms =
    [
        "Widget",
        "Element",
        "RenderObject",
        "SearchBar",
        "SearchAnchor",
        "Navigator",
        "Material",
        "TextField",
        "ThemeData",
        "Plumix",
    ];

    private readonly SearchController _controller = new();
    private readonly TextEditingController _standaloneController = new("Standalone");
    private SearchDelegate<string>? _legacyDelegate;
    private bool _enabled = true;
    private bool _useFullScreen;
    private bool _useThemeOverrides;
    private string _selected = "none";
    private string _status = "idle";

    public override void InitState()
    {
        _legacyDelegate = new TermSearchDelegate(SearchTerms, value => SetState(() =>
        {
            _selected = value;
            _status = $"legacy selected: {value}";
        }));
    }

    public override Widget Build(BuildContext context)
    {
        var baseTheme = Theme.Of(context);
        var pageTheme = baseTheme with
        {
            SearchBarTheme = _useThemeOverrides
                ? new SearchBarThemeData(
                    BackgroundColor: MaterialStateProperty<Color?>.All(Color.Parse("#FFEAF6F7")),
                    Elevation: MaterialStateProperty<double?>.All(1),
                    Shape: MaterialStateProperty<ShapeBorder?>.All(new RoundedRectangleBorder(borderRadius:
                        Plumix.Rendering.BorderRadius.Circular(18))),
                    Side: MaterialStateProperty<BorderSide?>.All(new BorderSide(Color.Parse("#FF00695C"))),
                    Padding: MaterialStateProperty<Thickness?>.All(new Thickness(12, 0)),
                    Constraints: new BoxConstraints(MinWidth: 280, MaxWidth: 520, MinHeight: 52),
                    HintStyle: MaterialStateProperty<TextStyle?>.All(new TextStyle(Color: Color.Parse("#FF00695C"))))
                : new SearchBarThemeData(),
            SearchViewTheme = _useThemeOverrides
                ? new SearchViewThemeData(
                    BackgroundColor: Color.Parse("#FFF5FBFA"),
                    Elevation: 0,
                    Shape: new RoundedRectangleBorder(
                        new BorderSide(Color.Parse("#FF80CBC4")), Plumix.Rendering.BorderRadius.Circular(22)),
                    HeaderHeight: 64,
                    BarPadding: new Thickness(12, 0),
                    DividerColor: Color.Parse("#FF80CBC4"),
                    Constraints: new BoxConstraints(MinWidth: 360, MinHeight: 260, MaxWidth: 560, MaxHeight: 420),
                    Padding: new Thickness(16))
                : new SearchViewThemeData()
        };

        return new Theme(
            data: pageTheme,
            child: new SingleChildScrollView(
                child: new Column(
                    crossAxisAlignment: CrossAxisAlignment.Stretch,
                    spacing: 12,
                    children:
                    [
                        new Text("SearchBar + SearchAnchor + SearchDelegate", fontSize: 20, color: Colors.Black),
                        new Text(
                            "Controller-backed search plus a legacy route with search-keyboard metadata "
                            + "and fade animation.",
                            fontSize: 14,
                            color: Color.Parse("#8A000000")),
                        new Row(
                            spacing: 8,
                            children:
                            [
                                ControlButton(_enabled ? "Enabled" : "Disabled", () => SetState(() => _enabled = !_enabled)),
                                ControlButton(_useFullScreen ? "Full screen" : "Docked view", () => SetState(() => _useFullScreen = !_useFullScreen)),
                                ControlButton(_useThemeOverrides ? "Theme on" : "Theme off", () => SetState(() => _useThemeOverrides = !_useThemeOverrides)),
                                ControlButton("Legacy route", () => OpenLegacySearch(context)),
                            ]),
                        new Row(
                            spacing: 8,
                            children:
                            [
                                ControlButton("Open", () => _controller.OpenView()),
                                ControlButton("Clear", () => SetState(() =>
                                {
                                    _controller.Clear();
                                    _selected = "none";
                                    _status = "cleared";
                                })),
                            ]),
                        new Align(
                            alignment: Alignment.CenterLeft,
                            child: SearchAnchor.Bar(
                                searchController: _controller,
                                barHintText: "Search framework terms",
                                viewHintText: "Type a Plumix concept",
                                isFullScreen: _useFullScreen,
                                enabled: _enabled,
                                shrinkWrap: true,
                                constraints: new BoxConstraints(MinWidth: 320, MaxWidth: 560, MinHeight: 56),
                                viewConstraints: new BoxConstraints(MinWidth: 360, MinHeight: 260, MaxWidth: 560, MaxHeight: 420),
                                barTrailing:
                                [
                                    new IconButton(
                                        icon: new Icon(Icons.Clear),
                                        onPressed: () => SetState(() =>
                                        {
                                            _controller.Clear();
                                            _status = "bar cleared";
                                        }))
                                ],
                                onOpen: () => SetState(() => _status = "opened"),
                                onClose: () => SetState(() => _status = "closed"),
                                onChanged: value => SetState(() => _status = $"changed: {FormatEmpty(value)}"),
                                onSubmitted: value => SetState(() => _status = $"submitted: {FormatEmpty(value)}"),
                                suggestionsBuilder: BuildSuggestions)),
                        new Text($"Selected: {_selected}", fontSize: 13),
                        new Text($"Controller text: {FormatEmpty(_controller.Text)}", fontSize: 13),
                        new Text($"Status: {_status}", fontSize: 13),
                        new Divider(),
                        new Text("Standalone SearchBar", fontSize: 18, color: Colors.Black),
                        new SearchBar(
                            controller: _standaloneController,
                            hintText: "Filter inside a page",
                            leading: new Icon(Icons.Search),
                            trailing:
                            [
                                new IconButton(
                                    icon: new Icon(Icons.Clear),
                                    onPressed: () => SetState(() => _standaloneController.Clear()))
                            ],
                            onTap: () => SetState(() => _status = "standalone tapped"),
                            onChanged: value => SetState(() => _status = $"standalone: {FormatEmpty(value)}"),
                            onSubmitted: value => SetState(() => _status = $"standalone submitted: {FormatEmpty(value)}")),
                    ])));
    }

    public override void Dispose()
    {
        _controller.Dispose();
        _standaloneController.Dispose();
        _legacyDelegate?.Dispose();
    }

    private IReadOnlyList<Widget> BuildSuggestions(BuildContext context, SearchController controller)
    {
        string query = controller.Text.Trim();
        var suggestions = new List<Widget>();
        foreach (string term in SearchTerms)
        {
            if (query.Length > 0 && !term.Contains(query, StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            suggestions.Add(new ListTile(
                leading: new Icon(Icons.Search),
                title: new Text(term),
                subtitle: new Text($"Select {term}"),
                onTap: () => SetState(() =>
                {
                    _selected = term;
                    _status = $"selected: {term}";
                    controller.CloseView(term);
                })));
        }

        if (suggestions.Count == 0)
        {
            suggestions.Add(new Padding(
                new Thickness(16),
                new Text($"No results for {query}", fontSize: 13, color: Colors.DimGray)));
        }

        return suggestions;
    }

    private static Widget ControlButton(string label, Action action)
    {
        return new TextButton(new Text(label, fontSize: 12), action);
    }

    private static string FormatEmpty(string value)
    {
        return string.IsNullOrEmpty(value) ? "empty" : value;
    }

    private async void OpenLegacySearch(BuildContext context)
    {
        var searchDelegate = _legacyDelegate;
        if (searchDelegate is null || searchDelegate.IsActive)
        {
            return;
        }

        _status = "legacy opened";
        string? result = await MaterialSearch.ShowSearch(context, searchDelegate, query: _controller.Text);
        if (result is not null && Mounted)
        {
            SetState(() => _status = $"legacy closed: {result}");
        }
    }

    private sealed class TermSearchDelegate : SearchDelegate<string>
    {
        private readonly IReadOnlyList<string> _terms;
        private readonly Action<string> _onSelected;

        public TermSearchDelegate(IReadOnlyList<string> terms, Action<string> onSelected) : base(
            searchFieldLabel: "Search framework terms",
            keyboardType: TextInputType.Text,
            textInputAction: TextInputAction.Search,
            autocorrect: false,
            enableSuggestions: true)
        {
            _terms = terms;
            _onSelected = onSelected;
        }

        public override Widget BuildSuggestions(BuildContext context) => BuildTerms(context, "Suggestions");

        public override Widget BuildResults(BuildContext context) => BuildTerms(context, "Results");

        public override Widget? BuildLeading(BuildContext context) => new FadeTransition(
            opacity: TransitionAnimation,
            child: new BackButton(onPressed: () => Navigator.MaybePop(context)));

        public override IReadOnlyList<Widget>? BuildActions(BuildContext context) =>
        [
            new Tooltip(
                message: MaterialLocalizations.Of(context).ClearButtonTooltip,
                child: new IconButton(
                    icon: new Icon(Icons.Clear),
                    onPressed: () =>
                    {
                        Query = string.Empty;
                        ShowSuggestions();
                    })),
        ];

        private Widget BuildTerms(BuildContext context, string label)
        {
            string query = Query.Trim();
            var children = new List<Widget> { new Text($"{label} for {FormatEmpty(query)}", fontSize: 13) };
            foreach (string term in _terms)
            {
                if (query.Length > 0 && !term.Contains(query, StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                children.Add(new ListTile(
                    leading: new Icon(Icons.Search),
                    title: new Text(term),
                    onTap: () =>
                    {
                        if (label == "Suggestions")
                        {
                            Query = term;
                            ShowResults();
                        }
                        else
                        {
                            _onSelected(term);
                            Close(context, term);
                        }
                    }));
            }

            return new ListView(children: children);
        }
    }
}

File: dart_sample/lib/demos/material/search_demo_page.dart
import 'package:flutter/material.dart';

class SearchDemoPage extends StatefulWidget {
  const SearchDemoPage({super.key});

  @override
  State<SearchDemoPage> createState() => _SearchDemoPageState();
}

class _SearchDemoPageState extends State<SearchDemoPage> {
  static const List<String> _searchTerms = <String>[
    'Widget',
    'Element',
    'RenderObject',
    'SearchBar',
    'SearchAnchor',
    'Navigator',
    'Material',
    'TextField',
    'ThemeData',
    'Plumix',
  ];

  final SearchController _controller = SearchController();
  final TextEditingController _standaloneController = TextEditingController(
    text: 'Standalone',
  );
  bool _enabled = true;
  bool _useFullScreen = false;
  bool _useThemeOverrides = false;
  String _selected = 'none';
  String _status = 'idle';

  @override
  Widget build(BuildContext context) {
    final ThemeData baseTheme = Theme.of(context);
    final ThemeData pageTheme = baseTheme.copyWith(
      searchBarTheme: _useThemeOverrides
          ? SearchBarThemeData(
              backgroundColor: const WidgetStatePropertyAll<Color?>(
                Color(0xFFEAF6F7),
              ),
              elevation: const WidgetStatePropertyAll<double?>(1),
              shape: WidgetStatePropertyAll<OutlinedBorder?>(
                RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(18),
                  side: const BorderSide(color: Color(0xFF00695C)),
                ),
              ),
              padding: const WidgetStatePropertyAll<EdgeInsetsGeometry?>(
                EdgeInsets.symmetric(horizontal: 12),
              ),
              constraints: const BoxConstraints(
                minWidth: 280,
                maxWidth: 520,
                minHeight: 52,
              ),
              hintStyle: const WidgetStatePropertyAll<TextStyle?>(
                TextStyle(color: Color(0xFF00695C)),
              ),
            )
          : const SearchBarThemeData(),
      searchViewTheme: _useThemeOverrides
          ? SearchViewThemeData(
              backgroundColor: const Color(0xFFF5FBFA),
              elevation: 0,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(22),
                side: const BorderSide(color: Color(0xFF80CBC4)),
              ),
              headerHeight: 64,
              barPadding: const EdgeInsets.symmetric(horizontal: 12),
              dividerColor: const Color(0xFF80CBC4),
              constraints: const BoxConstraints(
                minWidth: 360,
                minHeight: 260,
                maxWidth: 560,
                maxHeight: 420,
              ),
              padding: const EdgeInsets.all(16),
            )
          : const SearchViewThemeData(),
    );

    return Theme(
      data: pageTheme,
      child: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          spacing: 12,
          children: <Widget>[
            const Text(
              'SearchBar + SearchAnchor + SearchDelegate',
              style: TextStyle(fontSize: 20, color: Colors.black),
            ),
            const Text(
              'Controller-backed search view with suggestions, open/close callbacks, M3 defaults, and theme precedence probes.',
              style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
            ),
            Row(
              spacing: 8,
              children: <Widget>[
                _controlButton(
                  _enabled ? 'Enabled' : 'Disabled',
                  () => setState(() => _enabled = !_enabled),
                ),
                _controlButton(
                  _useFullScreen ? 'Full screen' : 'Docked view',
                  () => setState(() => _useFullScreen = !_useFullScreen),
                ),
                _controlButton(
                  _useThemeOverrides ? 'Theme on' : 'Theme off',
                  () =>
                      setState(() => _useThemeOverrides = !_useThemeOverrides),
                ),
                _controlButton(
                  'Legacy route',
                  () => _openLegacySearch(context),
                ),
              ],
            ),
            Row(
              spacing: 8,
              children: <Widget>[
                _controlButton('Open', () => _controller.openView()),
                _controlButton(
                  'Clear',
                  () => setState(() {
                    _controller.clear();
                    _selected = 'none';
                    _status = 'cleared';
                  }),
                ),
              ],
            ),
            Align(
              alignment: Alignment.centerLeft,
              child: SearchAnchor.bar(
                searchController: _controller,
                barHintText: 'Search framework terms',
                viewHintText: 'Type a Plumix concept',
                isFullScreen: _useFullScreen,
                enabled: _enabled,
                shrinkWrap: true,
                constraints: const BoxConstraints(
                  minWidth: 320,
                  maxWidth: 560,
                  minHeight: 56,
                ),
                viewConstraints: const BoxConstraints(
                  minWidth: 360,
                  minHeight: 260,
                  maxWidth: 560,
                  maxHeight: 420,
                ),
                barTrailing: <Widget>[
                  IconButton(
                    icon: const Icon(Icons.clear),
                    onPressed: () => setState(() {
                      _controller.clear();
                      _status = 'bar cleared';
                    }),
                  ),
                ],
                onOpen: () => setState(() => _status = 'opened'),
                onClose: () => setState(() => _status = 'closed'),
                onChanged: (String value) =>
                    setState(() => _status = 'changed: ${_formatEmpty(value)}'),
                onSubmitted: (String value) => setState(
                  () => _status = 'submitted: ${_formatEmpty(value)}',
                ),
                suggestionsBuilder: _buildSuggestions,
              ),
            ),
            Text('Selected: $_selected', style: const TextStyle(fontSize: 13)),
            Text(
              'Controller text: ${_formatEmpty(_controller.text)}',
              style: const TextStyle(fontSize: 13),
            ),
            Text('Status: $_status', style: const TextStyle(fontSize: 13)),
            const Divider(),
            const Text(
              'Standalone SearchBar',
              style: TextStyle(fontSize: 18, color: Colors.black),
            ),
            SearchBar(
              controller: _standaloneController,
              hintText: 'Filter inside a page',
              leading: const Icon(Icons.search),
              trailing: <Widget>[
                IconButton(
                  icon: const Icon(Icons.clear),
                  onPressed: () => setState(_standaloneController.clear),
                ),
              ],
              onTap: () => setState(() => _status = 'standalone tapped'),
              onChanged: (String value) => setState(
                () => _status = 'standalone: ${_formatEmpty(value)}',
              ),
              onSubmitted: (String value) => setState(
                () => _status = 'standalone submitted: ${_formatEmpty(value)}',
              ),
            ),
          ],
        ),
      ),
    );
  }

  Iterable<Widget> _buildSuggestions(
    BuildContext context,
    SearchController controller,
  ) {
    final String query = controller.text.trim();
    final List<Widget> suggestions = <Widget>[];
    for (final String term in _searchTerms) {
      if (query.isNotEmpty &&
          !term.toLowerCase().contains(query.toLowerCase())) {
        continue;
      }

      suggestions.add(
        ListTile(
          leading: const Icon(Icons.search),
          title: Text(term),
          subtitle: Text('Select $term'),
          onTap: () => setState(() {
            _selected = term;
            _status = 'selected: $term';
            controller.closeView(term);
          }),
        ),
      );
    }

    if (suggestions.isEmpty) {
      suggestions.add(
        Padding(
          padding: const EdgeInsets.all(16),
          child: Text(
            'No results for $query',
            style: const TextStyle(fontSize: 13, color: Colors.grey),
          ),
        ),
      );
    }

    return suggestions;
  }

  @override
  void dispose() {
    _controller.dispose();
    _standaloneController.dispose();
    super.dispose();
  }

  Widget _controlButton(String label, VoidCallback onPressed) {
    return TextButton(onPressed: onPressed, child: Text(label));
  }

  static String _formatEmpty(String value) {
    return value.isEmpty ? 'empty' : value;
  }

  Future<void> _openLegacySearch(BuildContext context) async {
    final String? result = await showSearch<String>(
      context: context,
      query: _controller.text,
      delegate: _TermSearchDelegate(_searchTerms, (String value) {
        setState(() {
          _selected = value;
          _status = 'legacy selected: $value';
        });
      }),
    );
    if (result != null && mounted) {
      setState(() => _status = 'legacy closed: $result');
    }
  }
}

class _TermSearchDelegate extends SearchDelegate<String> {
  _TermSearchDelegate(this._terms, this._onSelected)
    : super(searchFieldLabel: 'Search framework terms');

  final List<String> _terms;
  final ValueChanged<String> _onSelected;

  @override
  List<Widget>? buildActions(BuildContext context) {
    return <Widget>[
      IconButton(
        icon: const Icon(Icons.clear),
        tooltip: MaterialLocalizations.of(context).clearButtonTooltip,
        onPressed: () {
          query = '';
          showSuggestions(context);
        },
      ),
    ];
  }

  @override
  Widget? buildLeading(BuildContext context) {
    return BackButton(onPressed: () => close(context, null));
  }

  @override
  Widget buildResults(BuildContext context) => _buildTerms(context, 'Results');

  @override
  Widget buildSuggestions(BuildContext context) =>
      _buildTerms(context, 'Suggestions');

  Widget _buildTerms(BuildContext context, String label) {
    final String normalizedQuery = query.trim();
    return ListView(
      children: <Widget>[
        Text(
          '$label for ${_SearchDemoPageState._formatEmpty(normalizedQuery)}',
        ),
        for (final String term in _terms)
          if (normalizedQuery.isEmpty ||
              term.toLowerCase().contains(normalizedQuery.toLowerCase()))
            ListTile(
              leading: const Icon(Icons.search),
              title: Text(term),
              onTap: () {
                if (label == 'Suggestions') {
                  query = term;
                  showResults(context);
                } else {
                  _onSelected(term);
                  close(context, term);
                }
              },
            ),
      ],
    );
  }
}
